tetramer_validator/validate.py

```diff
@@ -11,147 +11,332 @@
 PTM_file = "data/PTM_list.tsv"
 PTM_file = path.join(here, PTM_file)
 
+errors = []
 with open(PTM_file) as fh_1:
     reader = csv.DictReader(fh_1, delimiter="\t")
     PTM_display = [name["display_name"] for name in reader]
 
 
-def validate(pep_seq, mhc_name=None, mod_type=None, mod_pos=None):
+def generate_problem_table(level, rule_name, value, instructions, fix=None):
+    global errors
+    errors.append(locals())
+    print("generae")
+    print(errors)
+
+def validate(pep_seq, mhc_name, mod_type=None, mod_pos=None):
+    """Main validate function."""
     args = locals()
-    null_str = "NULL"
-    if null_str in args.values():
-        return "NULL value entered. If there is no need for a particular value, please leave blank."
-    if "" == args["pep_seq"]:
-        return "Incorrect number of arguments: Peptide sequence is required"
-    if pep_seq != "" and mhc_name is None and mod_type is None and mod_pos is None:
-        return (
-            "Incorrect number of arguments: Please enter modification information "
-            "and/or MHC molecule information."
+    global errors
+    errors = []
+    properNumArguments(args)
+    null_input_check(args)
+
+    if errors:
+        return errors
+
+    if mod_pos and mod_type:
+        validate_peptide(pep_seq, mod_type=mod_type, mod_pos=mod_pos)
+    print(errors)
+    validate_mhc_name(mhc_name)
+    print("validate")
+    print(errors)
+    return errors
+
+
+def properNumArguments(args):
+    """Checks for proper number and combination of arguments for validate function"""
+
+    incorrect_num_str = "IncorrectNumArgs"
+
+    if not args["pep_seq"]:
+        generate_problem_table(
+            level="error",
+            rule_name=incorrect_num_str,
```

Comment:
The point of the rule name is to distinguish different problems. I expect that every call to `generate_problem_table` is a different problem, and should have a different name.

However the problems are named, they should follow the same pattern, starting with the general and becoming more specific. This makes it easier to sort by problem name.

Since you know that you want to highlight a specific input in the HTML form, it would be good to include that field name here, e.g. `field=pep_seq`.

Reply:
Partly resolved by commit ff62bf1 and commit 29a289a